src/regexp.toit

```diff
@@ -1058,70 +1084,106 @@ class CharClass extends MiniExpAst:
+  fix_ranges ranges/List/*<Range_>*/ case_sensitive/bool -> none:
+    // There's a lot of punctuation and no case-sensitive characters before the
+    // Unicode 'A' code point.
+    if not case_sensitive and (ranges.any: it.to >= 'A'):
+      old_size := ranges.size
+      range_start := -1
+      range_end := -1
+      for i := 0; i < old_size; i++:
+        range := ranges[i]
+        for j := range.from; j <= range.to; j++:
+          equivalence := case.reg_exp_equivalence_class j
+          if equivalence and equivalence.size != 1:
+            if range_start >= 0 and equivalence.size == 2 and equivalence.contains range_end + 1 and j != range_end:
+              // There's just one case-equivalent character and it's the one we
+              // were expecting.
+              range_end++
+            else:
+              assert: equivalence.contains j
+              // Flush the existing range of equivalents that we were
+              // constructing.
+              if range_start >= 0:
+                ranges.add (Range_ range_start range_end)
+                range_start = -1
+              if equivalence.size > 2:
+                // If there are > 2 equivalents, just add them all to the list
+                // of ranges.
+                equivalence.do: | equivalent |
+                  ranges.add (Range_ equivalent equivalent)
+              else:
+                // Since there are two equivalents, and one of them is the
+                // character we already have in a range, start constructing a
+                // new range of equivalents.
+                equivalence.do: | equivalent |
+                  if equivalent != j:
+                    range_start = range_end = equivalent
+        // At the end, flush the equivalent range we were constructing.
+        if range_start >= 0:
+          ranges.add (Range_ range_start range_end)
+          range_start = -1
+
+    // Even if there is no case-independence work to do we still want to
+    // consolidate ranges.
+    ranges.sort --in_place: | a b | a.from - b.from
+
+    // Now that they are sorted, check if it's worth merging
+    // adjacent/overlapping ranges.
+    for i := 1; i < ranges.size; i++:
+      if ranges[i - 1].to + 1 >= ranges[i].from:
+        ranges = merge_adjacent_ranges_ ranges
+        break
+
+  // Merge adjacent/overlapping ranges.
+  merge_adjacent_ranges_ ranges/List -> List:
+    last := ranges[0]
+    new_ranges := [last]
+    ranges.do: | range |
+      if last.to + 1 >= range.from:
+        last.to = max last.to range.to
       else:
-        if start != -1:
-          ranges.add 'A' + start
-          ranges.add 'A' + i - 1
-          ranges.add 'a' + start
-          ranges.add 'a' + i - 1
-          start = -1
-    // TODO(erikcorry): Sort and merge ranges.
-    return ranges
+        last = range
+        new_ranges.add last
+    return new_ranges
+
+  case_expand compiler/MiniExpCompiler -> MiniExpAst:
+    ranges := ranges_.copy
```

Comment:
Since you make a copy here anyway, might as well do the copy in the function that modifies the list.